File: problems/698/solution.go
package main
func canPartitionKSubsets(nums []int, k int) bool {
    sum, max := 0, 0
    
    for _, num := range nums {
        sum += num
        
        if num > max {
            max = num
        }
    }
    
    if sum % k != 0 || max > sum / k {
        return false
    }
    
    sum /= k
    
    return backtrack(nums, k, sum, 0, 0)
}

func backtrack(nums []int, k, sum, curSum, i int) bool {
    if k == 1 {
        return true
    }
    
    if sum == curSum {
        return backtrack(nums, k - 1, sum, 0, 0)
    }
    
    for j := i; j < len(nums); j++ {
        if nums[j] > 0 && curSum + nums[j] <= sum {
            nums[j] *= -1
            
            if (backtrack(nums, k, sum, curSum + nums[j] * -1, j + 1)) {
                return true
            }
            
            nums[j] *= -1
        }
    }
    
    return false
}

File: problems/698/solution.cpp
class Solution {
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        int sum = accumulate(begin(nums), end(nums), 0);
        
        if (sum % k != 0) {
            return false;
        }
        
        sum /= k;
        
        if (*max_element(begin(nums), end(nums)) > sum) {
            return false;
        }
        
        return backtrack(nums, k, sum, 0, 0);
    }
    
    bool backtrack(vector<int>& nums, int k, const int sum, int curSum, int i) {
        if (k == 1) {
            return true;
        }
        
        if (sum == curSum) {
            return backtrack(nums, k - 1, sum, 0, 0);
        }
        
        for (int j = i; j < nums.size(); j++) {
            if (nums[j] > 0 and curSum + nums[j] <= sum) {
                nums[j] *= -1;
                
                if (backtrack(nums, k, sum, curSum + nums[j] * -1, i + 1)) {
                    return true;
                }
                
                nums[j] *= -1;
            }
        }
        
        return false;
    }
};